Add markdown rendering for text content blocks

LukasKalbertodt opened the pull request:
See commit messages for more details. Especially the third one:

> This uses the package `react-markdown`. Unfortunately, this adds 32kB
> to our gzipped bundle size bringing it from 174kB to 206kB. I only
> found one other library that could be used and was only 5kB. However,
> it was not as popular and more importantly: it didn't have as good of
> a default behavior regarding raw HTML and didn't allow to customize
> which elements are allowed.
> 
> The markdown that's being rendered here is a subset of CommonMark. I
> did not want to allow everything right away for several reasons. Most
> notably missing: headings and images. See code comments for more
> information.
> 
> Unfortunately also missing are autolinks and strike-through. The former
> is something I am used to with GitHub flavored Markdown. Just writing
> `github.com` or `https://foo.de` would create a link. Not so in this
> PR, where everything has to be explicitly linked with `[]()` or
> "explicitly autolinked" with `<https://foo.de>`.
> 
> Raw HTML is escaped.

You can review this PR commit by commit which also allows you to ignore the commit where I just split one file into multiple.

Here is a picture of all Markdown things in action:

![murkdown](https://user-images.githubusercontent.com/7419664/117344585-d1f0fb00-aea5-11eb-9c8a-b2da8a472826.png)

CC https://github.com/elan-ev/tobira/issues/168

LukasKalbertodt commented:
Open questions:

- Should we immediately allow headers? And if so, we need to "shift their level down", i.e. `# foo` cannot become `h1` because we use `h1` outside of textblocks for other things. We are also already using `h2` somewhere in our code. So should `# foo` get `h3`?
- Should we allow images already despite having no way to upload them? 
- Are all other Markdown things we allow here fine? Or do we want to now allow some of those?

JulianKniephoff commented on frontend/src/layout/NavMain.tsx:
```diff
@@ -46,7 +46,7 @@ export const NavMain: React.FC<Props> = ({ title, breadcrumbs, children, ...navP
     }}>
         <div css={{ gridArea: "breadcrumbs" }}>{breadcrumbs}</div>
-        <h1 css={{ gridArea: "title", margin: "12px 0" }}>{title}</h1>
+        {title !== undefined && <h1 css={{ gridArea: "title", margin: "12px 0" }}>{title}</h1>}
```
```suggestion
        {title && <h1 css={{ gridArea: "title", margin: "12px 0" }}>{title}</h1>}
```

would be shorter and would also catch `''`, or do we want an empty header for an empty title? :thinking:

LukasKalbertodt replied:
That's what I wrote first, but got this:

```
error  Unexpected nullable string value in conditional. Please handle the nullish/empty cases explicitly  @typescript-eslint/strict-boolean-expressions
```

So... disable that lint?

JulianKniephoff replied:
Hmm, good point. No, I would leave it, that's actually useful sometimes. (I ran into `0` being falsy quite a few times already ...)

I'm still disappointed that there is no `x` in JS that completes this analogy: `??` : `||` :: `x` : `&&`. :disappointed:

JulianKniephoff commented on frontend/src/ui/Blocks.tsx:
```diff
@@ -1,4 +1,6 @@
 import React from "react";
+import ReactMarkdown from "react-markdown";
+import { Components } from "react-markdown/src/ast-to-react";
```
Importing from `src` seems a bit weird? Is this intended by the authors?

JulianKniephoff commented on frontend/src/ui/Blocks.tsx:
```diff
@@ -57,14 +59,60 @@ type TextProps = {
+// We override some components emitted by the Markdown parser to add CSS.
+const MARKDOWN_COMPONENTS: Components = {
+    p: ({ node: _, ...props }) => <p css={{ margin: "16px 0" }} {...props} />,
+    blockquote: ({ node, ...props }) => <blockquote
```
`node` is ignored in both cases, but it's notated differently.

JulianKniephoff commented on frontend/src/ui/Blocks.tsx:
```diff
@@ -57,14 +59,60 @@ type TextProps = {
+        {...props}
+    />,
+    code: ({ node, className, inline, ...props }) => <code
+        css={inline === true && {
```
is `inline: boolean`? or `boolean?`? In any case, `inline === true` should be equivalent to `inline`:

```suggestion
        css={inline && {
```

LukasKalbertodt replied:
Similar to above, with this I get:

```
error  Unexpected nullable boolean value in conditional. Please handle the nullish case explicitly  @typescript-eslint/strict-boolean-expressions
```

JulianKniephoff replied:
Oof, that sucks even more. But it makes sense. We could apparently [configure the rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md) to ignore certain types, but I honestly wouldn't, I think. :thinking:

JulianKniephoff commented on frontend/src/ui/Blocks.tsx:
```diff
@@ -57,14 +59,60 @@ type TextProps = {
+
+// We override some components emitted by the Markdown parser to add CSS.
+const MARKDOWN_COMPONENTS: Components = {
+    p: ({ node: _, ...props }) => <p css={{ margin: "16px 0" }} {...props} />,
```
Btw. I think I slightly prefer this way of ignoring props. I also wonder why we can't just pass `node` along as well, though. :thinking:

JulianKniephoff replied:
And that also means I'm not sure about the added lint option, for the record. x)

LukasKalbertodt replied:
So, this only works for one parameter. `{ node: _, className: _, ...rest }` does not work because of using the same name twice. So that would mean `node: _node` and `className: _className`. Mhhh. Is that still better in your opinion?

And yeah dunno what would happen if we pass node along. Like... we probably don't want to pass random props to components right? What if those components suddenly start using them?

JulianKniephoff replied:
Hmm, yeah, those are good points. :thinking: Aight, I give up. :shrug:

JulianKniephoff reviewed (changes requested):
Very cool! :D As usual only a few minor nitpicks.

JulianKniephoff commented:
> * Should we immediately allow headers? And if so, we need to "shift their level down", i.e. `# foo` cannot become `h1` because we use `h1` outside of textblocks for other things. We are also already using `h2` somewhere in our code. So should `# foo` get `h3`?

HTML allows for independent hierarchies for different sections. However, there are accessibility concerns to be considered here since screen readers use headlines to build tables of contents sometimes, apparently. I don't know how these two facts interact. Maybe we could ask Alireza about this?

> * Should we allow images already despite having no way to upload them?

We also don't have a way to input text, yet. :wink: I don't see a reason not to allow them; you can always hotlink URLs, or put something up in on nginx yourself. :shrug: 

> * Are all other Markdown things we allow here fine? Or do we want to now allow some of those?

I think it's fine. I see tables missing as well, though. I could imagine academic users wanting those. :thinking:

LukasKalbertodt commented:
> We also don't have a way to input text, yet. wink I don't see a reason not to allow them; you can always hotlink URLs, or put something up in on nginx yourself. shrug

Yeah I guess, why not.

> I think it's fine. I see tables missing as well, though. I could imagine academic users wanting those. thinking

I guess we can start slowly. Tables, strikethrough and auto linked URLs are added by GH-flavored markdown. We might want to add it, but I would suggest first merging this basic support. Stuff can always be added later.

---

> HTML allows for independent hierarchies for different sections. However, there are accessibility concerns to be considered here since screen readers use headlines to build tables of contents sometimes, apparently. I don't know how these two facts interact. Maybe we could ask Alireza about this?

I remember fairly well how he told us that it's bad to not use headers in strictly descending order, so to speak. Apparently GitHub renders `# foo` as a proper `h1`, but I'm not so sure this is a good idea.  From [this article](https://www.contentkingapp.com/academy/headings/):

> It is recommended to use only one H1 heading per page. Multiple H1 headings per page were introduced with HTML5, but browsers didn’t adopt the practice. Since HTML5.1, there is no support for multiple H1 headings.
>
> [...]
>
> When the HTML5 specification was first introduced, support for multiple H1 headings per page (each in their own section-element) was included. However, since browsers didn’t adopt this recommendation, support for this was subsequently dropped when the HTML5.1 standard was published. For more information, see this W3C thread (`https://github.com/w3c/html/issues/169`).

[MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements) also states:

> - Avoid skipping heading levels: always start from `<h1>`, followed by `<h2>` and so on.
> - Use only one `<h1>` per page or view. It should concisely describe the overall purpose of the content.
> - Using more than one `<h1>` will not result in an error, but is not considered a best practice. Using only one `<h1>` is beneficial for screenreader users, and SEO.
> - While HTML5 allows a `<h1>` per sectioning element, it is not considered best practice, and may subvert the expectations of how screen reader users navigate.

There is also [this part on accessibility on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#accessibility_concerns) and [this article](https://www.w3.org/WAI/tutorials/page-structure/headings/). What I don't understand in those is what they mean by "nesting", "subsection" and "closing subsections". I'm not sure if they talk about implicitly opened/closed sections or whether they say one should actually use `<section>` whenever using a heading?

LukasKalbertodt commented:
So I addressed your review comments, pushed a commit to enable images and create a new issue for the headings discussion: #210. With that, this should be ready to merge.
